### PR TITLE
MINOR: Add system test for optimization upgrades

### PR DESCRIPTION
**Pull request by bbejeck**
This is a new system test testing for optimizing an existing topology.  This test takes the following steps

1. Start a Kafka Streams application that uses a `selectKey` then performs 3 `groupByKey()` operations and 1 `join` creating four repartition topics
2. Verify all instances start and process data
3. Stop all instances and verify stopped
4. For each stopped instance update the config for  `TOPOLOGY_OPTIMIZATION` to `all` then restart the instance and verify the instance has started successfully also verifying Kafka Streams reduced the number of repartition topics from 4 to 1
5. Verify that each instance is processing data from the `aggregation`, `reduce`, and `join` operation
6. Stop all instances and verify the shut down is complete.

For testing I ran two passes of the system test with 25 repeats for a total of 50 test runs.  

All test runs passed
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    """
+    Test doing upgrades of a Kafka Streams application
+    that is un-optimized initially then optimized
+    """
+
+    input_topic = 'inputTopic'
+    aggregation_topic = 'aggregationTopic'
+    reduce_topic = 'reduceTopic'
+    join_topic = 'joinTopic'
+    operation_pattern = 'AGGREGATED\|REDUCED\|JOINED'
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py:35]: `monitor.wait_for` uses `grep` without the `E` switch so I need to escape the `|` in the pattern

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        # produce records continually during the test
+        self.producer.start()
+
+        # start all processors unoptimized
+        for processor in processors:
+            self.set_topics(processor)
+            processor.CLEAN_NODE_ENABLED = False
+            self.verify_running_repartition_topic_count(processor, 4)
+
+        self.verify_processing(processors, False)
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py]: With 5 sub-topologies (1 source input, 4 repartition topic inputs) and 6 partitions, we end up with 30 tasks.  So it's very probable that each instance only has 2 of the 3 tested aggregations i,e (`AGGREGATED`, `REDUCED`), (`AGGEGRATED`, `JOINED`), (`REDUCED`, `JOINED`) so I just verify that the `STDOUT` log has something from the pattern. The `False` parameter is used to signal verification with the `operation_pattern` as a whole.

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+        self.stop_processors(processors)
+
+        # start again with topology optimized
+        for processor in processors:
+            processor.OPTIMIZED_CONFIG = 'all'
+            self.verify_running_repartition_topic_count(processor, 1)
+
+        self.verify_processing(processors, True)
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py]: Now there are 2 sub-topologies (1 source input, 1 repartition topic inputs) and 6 partitions; we end up with 12 tasks. So each streams instance should have at least one task for `AGGREGATED`, `REDUCED`, and `JOINED` so we use the `True` parameter to indicate test for the existence of _**each**_  operation in the `STDOUT` of each streams instance.

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            processor.stop()
+            monitor.wait_until('OPTIMIZE_TEST Streams Stopped',
+                               timeout_sec=60,
+                               err_msg="'OPTIMIZE_TEST Streams Stopped' message" + str(processor.node.account))
+
+    def verify_processing(self, processors, verify_individual_operations):
+        for processor in processors:
+            if not self.all_source_subtopology_tasks(processor):
+                if verify_individual_operations:
+                    for operation in self.operation_pattern.split('\|'):
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py:116]: Testing for the existence of each term by itself in the `STDOUT` file

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                        self._do_verify(processor, operation)
+                else:
+                    self._do_verify(processor, self.operation_pattern)
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py]: Testing for the entire pattern

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        node = processor.node
+        with node.account.monitor_log(processor.STDOUT_FILE) as monitor:
+            processor.stop()
+            monitor.wait_until('OPTIMIZE_TEST Streams Stopped',
+                               timeout_sec=60,
+                               err_msg="'OPTIMIZE_TEST Streams Stopped' message" + str(processor.node.account))
+
+    def verify_processing(self, processors, verify_individual_operations):
+        for processor in processors:
+            if not self.all_source_subtopology_tasks(processor):
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py:114]: In a tiny percentage of test runs one streams instance ends up with all input source tasks, i.e. (`0_0`, `0_1`, `0_2`, `0_3`), so none of the expected operations are processed on that node.  So we check the task assignment and if it's all input source tasks we skip checking this node.

I added this check after noticing some test flakiness.

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                self.logger.info("Skipping processor %s with all source tasks" % processor.node.account)
+
+    def _do_verify(self, processor, pattern):
+        self.logger.info("Verifying %s processing pattern in STDOUT_FILE" % pattern)
+        with processor.node.account.monitor_log(processor.STDOUT_FILE) as monitor:
+            monitor.wait_until(pattern,
+                               timeout_sec=60,
+                               err_msg="Never saw processing of %s " % pattern + str(processor.node.account))
+
+    def all_source_subtopology_tasks(self, processor):
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py:130]: Checking the task assignment for this processor node.

**Comment by bbejeck** [comment]
ping @guozhangwang, @mjsax, and @vvcephei for reviews

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.regex.Pattern;
+
+import static java.time.Duration.ofMillis;
+
+public class StreamsOptimizedTest {
+
+
+    public static void main(final String[] args) throws Exception {
+        if (args.length < 1) {
+            System.err.println("StreamsUpgradeTest requires one argument (properties-file) but no provided: ");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: nit: `StreamsUpgradeTest` -> `StreamsOptimizedTest`

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        }
+        final String propFileName = args.length > 0 ? args[0] : null;
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: nit: simplify to `final String propFileName = args[0];`

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        final Properties streamsProperties = Utils.loadProps(propFileName);
+
+        System.out.println("StreamsTest instance started StreamsOptimizedTest");
+        System.out.println("props=" + streamsProperties);
+
+        final String inputTopic = (String) streamsProperties.remove("input.topic");
+        final String aggregationTopic = (String) streamsProperties.remove("aggregation.topic");
+        final String reduceTopic = (String) streamsProperties.remove("reduce.topic");
+        final String joinTopic = (String) streamsProperties.remove("join.topic");
+        final boolean debugInput = Boolean.valueOf((String) streamsProperties.remove("debug.input"));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: Add `null` check for each?

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+        final Function<String, String> keyFunction = s -> Integer.toString(Integer.parseInt(s) % 9);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        final KStream<String, String> sourceStream = builder.stream(inputTopic, Consumed.with(Serdes.String(), Serdes.String()));
+
+        final KStream<String, String> mappedStream = sourceStream.selectKey((k, v) -> keyFunction.apply(v));
+
+        if (debugInput) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: Why this?

Also, there is a peek() in the other PR, that cannot be switched on/off. I like the idea, just want to get clarification to get a unique strategy we apply for system tests.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+            mappedStream.peek((k, v) -> System.out.println(String.format("MAPPED key=%s value=%s", k, v)));
+        }
+
+        final KStream<String, Long> countStream = mappedStream.groupByKey().count(Materialized.with(Serdes.String(), Serdes.Long())).toStream();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: nit: line too long

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+        mappedStream.groupByKey().aggregate(initializer,
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: nit: move `initializer` to next line

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+            aggregator,
+            Materialized.with(Serdes.String(), Serdes.Integer()))
+            .toStream()
+            .peek((k, v) -> System.out.println(String.format("AGGREGATED key=%s value=%s", k, v)))
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:91]: This `peek()` is not guarded like the one above

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+            .to(aggregationTopic, Produced.with(Serdes.String(), Serdes.Integer()));
+
+
+        mappedStream.groupByKey()
+            .reduce(reducer, Materialized.with(Serdes.String(), Serdes.String()))
+            .toStream()
+            .peek((k, v) -> System.out.println(String.format("REDUCED key=%s value=%s", k, v)))
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:98]: as above

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+            .to(reduceTopic, Produced.with(Serdes.String(), Serdes.String()));
+
+        mappedStream.join(countStream, (v1, v2) -> v1 + ":" + v2.toString(),
+            JoinWindows.of(ofMillis(500)),
+            Joined.with(Serdes.String(), Serdes.String(), Serdes.Long()))
+            .peek((k, v) -> System.out.println(String.format("JOINED key=%s value=%s", k, v)))
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:104]: one more :)

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+        final KafkaStreams streams = new KafkaStreams(topology, config);
+
+
+        streams.setStateListener((oldState, newState) -> {
+            if (oldState == State.REBALANCING && newState == State.RUNNING) {
+                final int repartitionTopicCount = getCountOfRepartitionTopicsFound(topology.describe().toString(), repartitionTopicPattern);
+                System.out.println(String.format("REBALANCING -> RUNNING with REPARTITION TOPIC COUNT=%d", repartitionTopicCount));
+                final Set<ThreadMetadata> localThreadsMetadata = streams.localThreadsMetadata();
+                for (final ThreadMetadata threadMetadata : localThreadsMetadata) {
+                    System.out.println("TASK ASSIGNMENT -> " + threadMetadata.activeTasks());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: Why do we need this? This info is logged already

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+            if not self.all_source_subtopology_tasks(processor):
+                if verify_individual_operations:
+                    for operation in self.operation_pattern.split('\|'):
+                        self._do_verify(processor, operation)
+                else:
+                    self._do_verify(processor, self.operation_pattern)
+            else:
+                self.logger.info("Skipping processor %s with all source tasks" % processor.node.account)
+
+    def _do_verify(self, processor, pattern):
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_optimized_test.py]: nit: why starting the name with `_` ?

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+        self.logger.info("Verifying %s processing pattern in STDOUT_FILE" % pattern)
+        with processor.node.account.monitor_log(processor.STDOUT_FILE) as monitor:
+            monitor.wait_until(pattern,
+                               timeout_sec=60,
+                               err_msg="Never saw processing of %s " % pattern + str(processor.node.account))
+
+    def all_source_subtopology_tasks(self, processor):
+        index = 0
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_optimized_test.py]: Why is this called `index`? Seems like a `retry` to me? -- Also, why do we need a retry? Can you use a `monitor` instead of direct `ssh_capture` ?

**Review by mjsax — COMMENTED** [review]
Not sure, if I fully understand the test setup, ie, what Python part does. Will revisit tomorrow again.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            .reduce(reducer, Materialized.with(Serdes.String(), Serdes.String()))
+            .toStream()
+            .peek((k, v) -> System.out.println(String.format("REDUCED key=%s value=%s", k, v)))
+            .to(reduceTopic, Produced.with(Serdes.String(), Serdes.String()));
+
+        mappedStream.join(countStream, (v1, v2) -> v1 + ":" + v2.toString(),
+            JoinWindows.of(ofMillis(500)),
+            Joined.with(Serdes.String(), Serdes.String(), Serdes.Long()))
+            .peek((k, v) -> System.out.println(String.format("JOINED key=%s value=%s", k, v)))
+            .to(joinTopic, Produced.with(Serdes.String(), Serdes.String()));
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:105]: Since we set default serde as String, String already, do we still need the Produced / Consumed when constructing the topology?

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+        });
+
+    }
+
+    private static int getCountOfRepartitionTopicsFound(final String topologyString,
+                                                        final Pattern repartitionTopicPattern) {
+        final Matcher matcher = repartitionTopicPattern.matcher(topologyString);
+        final List<String> repartitionTopicsFound = new ArrayList<>();
+        while (matcher.find()) {
+            repartitionTopicsFound.add(matcher.group());
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: nit: I'd suggest print the list to sysout as well for debugging, since the number of matched ones may not be sufficient.

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 6 above)
Comment (guozhangwang) [reply tests/kafkatest/tests/streams/streams_optimized_test.py:114]: Hmm.. I'm wondering why it is still possible, as we should have achieved balance across sub-topologies right? Or are there any potential edge cases you are aware of while working on that PR @bbejeck ?

**Review by guozhangwang — COMMENTED** [review]
Thanks @bbejeck . I just have some minor comments.

One meta question is that why we can still observe imbalance that tasks of the same sub-topology are not distributed evenly, do you have any ideas?

**Comment by guozhangwang** [comment]
@bbejeck as a side note, we believe that when doing this upgrade there may be a minor amount of data loss as some repartition topics that gets merged may have data not processed yet. Could you add some logs to illustrate how much data in percentage may be incurred in lost (we do not make it as a verification phase, but just print it out like "produced XXX, consumed YYY, lost XXX - YYY".

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 8 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 9 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 10 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 11 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: I needed this `Sysout` for debugging when writing the test, but I don't need to verify the output, hence the guard, but since it's no longer needed I'll remove it.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 12 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 13 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 14 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:91]: It's not guarded as the output is needed for verification in the test.  The one above was strictly for debugging purposes, but I've taken it out.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 15 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:98]: It's not guarded as the output is needed for verification in the test.  The one above was strictly for debugging purposes, but I've taken it out.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 16 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:104]: It's not guarded as the output is needed for verification in the test.  The one above was strictly for debugging purposes, but I've taken it out.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 17 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: yep, taking out

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 18 above)
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_optimized_test.py]: trying to mark it as a private method, but thinking about it more, that doesn't make too much sense, I'll remove the `_`.

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 19 above)
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_optimized_test.py]: updated variable name.  

I'll try using `monitor` , but I went with `ssh_capture` as it allows me to set my own regex and `monitor.wait_for` uses `grep` without the `-E` flag so I wasn't sure I could be as specific as I needed.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 20 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:105]: It's a habit for me to always put those in.   If you want me to remove them I will.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 21 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java]: ack

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 6 above)
Comment (bbejeck) [reply tests/kafkatest/tests/streams/streams_optimized_test.py:114]: I think as long as we favor stickiness over load-balancing this is always a possibility.  One thing to note I only observed one instance getting all tasks from one sub-topology _**after**_ the first phase of the test meaning stickiness is a factor, and it seemed to be a tiny percentage.  I put the check in to eliminate test flakiness.

I have some additional thoughts on putting back the check we had in making sure that adding a task from the same sub-topology only happens when all clients are over-capacity.  But I'd like to do that in a separate PR and write an independent system test for that.

WDYT?

**Comment by bbejeck** [comment]
@guozhangwang @mjsax updated per comments

**Comment by bbejeck** [comment]
>@bbejeck as a side note, we believe that when doing this upgrade there may be a minor amount of data loss as some repartition topics that gets merged may have data not processed yet. Could you add some logs to illustrate how much data in percentage may be incurred in lost (we do not make it as a verification phase, but just print it out like "produced XXX, consumed YYY, lost XXX - YYY".

This could be a little tricky as I'm using the `VerifiableProducer` with no message limit.  So this PR does not drag on can I do a separate PR adding a separate test?

**Comment by bbejeck** [comment]
Kicked off [another system test with 25 repeats all passed ](http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-11-16--001.1542400787--bbejeck--MINOR_create_system_test_for_rolling_upgrade_with_optimization--e372f97/report.html)

File: tests/kafkatest/tests/streams/streams_optimized_test.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+        self.logger.info("Verifying %s processing pattern in STDOUT_FILE" % pattern)
+        with processor.node.account.monitor_log(processor.STDOUT_FILE) as monitor:
+            monitor.wait_until(pattern,
+                               timeout_sec=60,
+                               err_msg="Never saw processing of %s " % pattern + str(processor.node.account))
+
+    def all_source_subtopology_tasks(self, processor):
+        retries = 0
+        while retries < 5:
+            found = list(processor.node.account.ssh_capture("sed -n 's/.*current active tasks: \[\(\(0_[0-9], \)\{3\}0_[0-9]\)\].*/\1/p' %s" % processor.LOG_FILE, allow_fail=True))
```
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_optimized_test.py:133]: changed this as `current active tasks` always shows up in the log files, where `Committed active tasks` may or may not be in the log file.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java (hunk 20 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsOptimizedTest.java:105]: No need to remove, just curious if there are any issues I do not know about that enforces you to add it :)

**Comment by guozhangwang** [comment]
> This could be a little tricky as I'm using the `VerifiableProducer` with no message limit. So this PR does not drag on can I do a separate PR adding a separate test?

Sounds good.

**Comment by guozhangwang** [comment]
LGTM. @mjsax feel free to merge after you made another pass.

**Comment by guozhangwang** [comment]
> One meta question is that why we can still observe imbalance that tasks of the same sub-topology are not distributed evenly, do you have any ideas?

@bbejeck any ideas?

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 18 above)
Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_optimized_test.py]: Is this a naming convention? Or a Python feature (ie, does an starting `_` make a method private in Python?)

**Comment by bbejeck** [comment]
>One meta question is that why we can still observe imbalance that tasks of the same sub-topology are not distributed evenly, do you have any ideas?
   @bbejeck any ideas?

Here's my original response https://github.com/apache/kafka/pull/5912#discussion_r234244608

Thinking about this some more, I also think setting ` group.initial.rebalance.delay.ms` to a higher value will help mitigate the issue.  Right now if one client gets tasks assignments before the other clients, it creates the potential for task imbalance.  But if all the groups are present at the same time, we should see an improvement in task assignment balance.  I'll experiment with this system test and see if it makes a difference.

**Comment by guozhangwang** [comment]
> Thinking about this some more, I also think setting ` group.initial.rebalance.delay.ms` to a higher value will help mitigate the issue. Right now if one client gets tasks assignments before the other clients, it creates the potential for task imbalance. But if all the groups are present at the same time, we should see an improvement in task assignment balance. I'll experiment with this system test and see if it makes a difference.

Thanks for the explanation, and it makes sense to me that if the imbalance only happens on the second phase, then it is indeed possible because maybe not all instances participated in the first rebalance. I agree that `initial.rebalance.delay.ms` would help but since it is a broker-side config as we discussed it before it is not a very well-designed configuration, and we may consider deprecating it as we improve the rebalance protocol in the future. So I think we do not need to spend more time experimenting it (again, just to clarify I agree with you that it will mitigate the issue, just think we do not need to spend more time validating that on system test :)

**Comment by bbejeck** [comment]
retest this please

**Comment by mjsax** [comment]
retest this please

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 2 above)
Comment (vvcephei) [reply tests/kafkatest/tests/streams/streams_optimized_test.py]: Can we use named parameters (especially for the boolean param) for clarity?

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 3 above)
Comment (vvcephei) [reply tests/kafkatest/tests/streams/streams_optimized_test.py]: cool. (again, do you mind using named params?)

File: tests/kafkatest/tests/streams/streams_optimized_test.py (hunk 6 above)
Comment (vvcephei) [reply tests/kafkatest/tests/streams/streams_optimized_test.py:114]: Interesting! I didn't realize we attempt to balance subtopologies over the instances... Is this important for some reason?

**Comment by vvcephei** [comment]
Hey @bbejeck ,

Sorry my review is late, but since the tests are still failing, maybe I can sneak in two comments (above, about named parameters in python)?

Regardless, it LGTM. Thanks!

**Comment by bbejeck** [comment]
@vvcephei updated per comments and kicked off new system test https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2104/

EDIT: rebased from trunk as well

**Comment by guozhangwang** [comment]
System test succeeded, merging to trunk now.
